Bonsai.Design/ObjectTextVisualizer.cs

```diff
@@ -30,27 +31,90 @@ public class ObjectTextVisualizer : BufferedVisualizer
         /// <inheritdoc/>
         protected override void ShowBuffer(IList<Timestamped<object>> values)
         {
-            if (values.Count > 0)
+            if (values.Count == 0)
+                return;
+
+            var sb = new StringBuilder();
+
+            // Trim old values if bufferSize was reduced
+            while (buffer.Count >= bufferSize)
+                buffer.Dequeue();
+
+            // Add new values to the buffer (and only the ones which might appear)
+            for (int i = Math.Max(0, values.Count - bufferSize); i < values.Count; i++)
             {
-                base.ShowBuffer(values);
-                textBox.Text = string.Join(Environment.NewLine, buffer);
-                textPanel.Invalidate();
+                var rawText = values[i].Value?.ToString() ?? string.Empty;
+                sb.Clear();
+                sb.EnsureCapacity(rawText.Length);
+                MakeDisplayString(rawText, sb);
+                DisplayString(sb.ToString());
             }
+
+            // Update the visual representation of the buffer
+            Debug.Assert(buffer.Count <= bufferSize);
+            sb.Clear();
+            foreach (var line in buffer)
+            {
+                if (sb.Length > 0)
+                    sb.Append(Environment.NewLine);
+                sb.Append(line);
+            }
+            textBox.Text = sb.ToString();
+            textPanel.Invalidate();
         }
 
         /// <inheritdoc/>
         public override void Show(object value)
         {
-            value ??= string.Empty;
-            var text = value.ToString();
-            text = Regex.Replace(text, @"\r|\n", string.Empty);
-            buffer.Enqueue(text);
-            while (buffer.Count > bufferSize)
+            var rawText = value?.ToString() ?? string.Empty;
+            var stringBuilder = new StringBuilder(rawText.Length);
+            MakeDisplayString(rawText, stringBuilder);
+            DisplayString(stringBuilder.ToString());
+        }
+
+        /// <summary>Converts the specified string for display in this visualizer</summary>
+        /// <param name="rawText">The raw text to be converted</param>
+        /// <param name="stringBuilder">The string builder which receives the display string</param>
+        protected virtual void MakeDisplayString(string rawText, StringBuilder stringBuilder)
```

Comment:
```suggestion
        /// <summary>
        /// Appends the display text for the specified object to the text buffer.
        /// </summary>
        /// <param name="stringBuilder">The string builder which receives the display text.</param>
        /// <param name="value">The object for which to retrieve the display text.</param>
        protected virtual void AppendDisplayText(StringBuilder stringBuilder, object value)
```

Since we are making this part of the API, I was considering we could refactor the method name to the above for consistency with `StringBuilder` methods and also for similarity with [`GetDisplayText`](https://learn.microsoft.com/en-us/dotnet/api/system.componentmodel.design.collectioneditor.getdisplaytext) which we use elsewhere for similar purposes.

The idea behind passing here the actual `object` is to give more flexibility for future specialized text visualizers which might want to access specific instance fields or properties of the object directly, e.g. reflection, etc for debugging purposes. Looking at the code I was hoping it shouldn't impact performance too much since `StringBuilder` should resize efficiently in most cases.

Reply:
I wasn't super happy with the name so this is much better, thanks!

I actually just moved the `EnsureCapacity` call to inside `AppendDisplayText` since that's really where it belonged anyway, so it really only leads to an extra allocation in the `Show(object)` case which we don't really care about.